bottlecap/src/lifecycle/invocation/triggers/kinesis_event.rs

```diff
@@ -0,0 +1,254 @@
+#![allow(clippy::module_name_repetitions)]
+use base64::engine::general_purpose;
+use base64::Engine;
+use datadog_trace_protobuf::pb::Span;
+use serde::{Deserialize, Serialize};
+use serde_json::{from_slice, Value};
+use std::collections::HashMap;
+use tracing::debug;
+
+use crate::lifecycle::invocation::{
+    processor::S_TO_NS,
+    triggers::{Trigger, DATADOG_CARRIER_KEY, FUNCTION_TRIGGER_EVENT_SOURCE_TAG},
+};
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct KinesisEvent {
+    #[serde(rename = "Records")]
+    pub records: Vec<KinesisRecord>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct KinesisRecord {
+    #[serde(rename = "eventID")]
+    pub event_id: String,
+    #[serde(rename = "eventName")]
+    pub event_name: String,
+    #[serde(rename = "eventSourceARN")]
+    pub event_source_arn: String,
+    #[serde(rename = "eventVersion")]
+    pub event_version: String,
+    pub kinesis: KinesisEntity,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct KinesisEntity {
+    #[serde(rename = "approximateArrivalTimestamp")]
+    pub approximate_arrival_timestamp: f64,
+    #[serde(rename = "partitionKey")]
+    pub partition_key: String,
+    pub data: String,
+}
+
+impl Trigger for KinesisRecord {
+    fn new(payload: Value) -> Option<Self> {
+        let records = payload.get("Records").and_then(Value::as_array);
+        match records {
+            Some(records) => match serde_json::from_value::<KinesisRecord>(records[0].clone()) {
+                Ok(event) => Some(event),
+                Err(e) => {
+                    debug!("Failed to deserialize Kinesis Record: {e}");
+                    None
+                }
+            },
+            None => None,
+        }
+    }
+
+    fn is_match(payload: &Value) -> bool {
+        if let Some(first_record) = payload
+            .get("Records")
+            .and_then(Value::as_array)
+            .and_then(|r| r.first())
+            .take()
+        {
+            first_record.get("kinesis").is_some()
+        } else {
+            false
+        }
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
+    fn enrich_span(&self, span: &mut Span) {
+        let event_source_arn = &self.event_source_arn;
+        let split_source_arn: Vec<&str> = event_source_arn.split('/').collect();
+        let parsed_stream_name = split_source_arn.last().unwrap_or(&"");
```

Comment:
To not do the or `&""` in DyamoDB I was able to do the following:

```rs
 let table_name = self.event_source_arn.split('/').nth(1).unwrap_or_default();
 ```

In SQS I saw the following:

```rs
let resource = self
    .event_source_arn
    .clone()
    .split(':')
    .last()
    .unwrap_or_default()
    .to_string();
    ```